Src/Newtonsoft.Json.Bson/BsonDataReader.Async.cs

```diff
@@ -0,0 +1,1018 @@
+            if (Value?.ToString() == Serialization.JsonTypeReflector.TypePropertyName)
+            {
+                await ReaderReadAndAssertAsync(cancellationToken).ConfigureAwait(false);
+                if (Value != null && Value.ToString().StartsWith("System.Byte[]", StringComparison.Ordinal))
+                {
+                    await ReaderReadAndAssertAsync(cancellationToken).ConfigureAwait(false);
+                    if (Value?.ToString() == Serialization.JsonTypeReflector.ValuePropertyName)
+                    {
+                        return;
+                    }
+                }
+            }
+
+            throw ExceptionUtils.CreateJsonReaderException(this, "Error reading bytes. Unexpected token: {0}.".FormatWith(CultureInfo.InvariantCulture, JsonToken.StartObject));
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next JSON token from the source as a <see cref="T:System.Nullable`1" /> of <see cref="T:System.DateTime" />.
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="P:System.Threading.CancellationToken.None" />.</param>
+        /// <returns>A <see cref="T:System.Threading.Tasks.Task`1" /> that represents the asynchronous read. The <see cref="P:System.Threading.Tasks.Task`1.Result" />
+        /// property returns the <see cref="T:System.Nullable`1" /> of <see cref="T:System.DateTime" />. This result will be <c>null</c> at the end of an array.</returns>
+        /// <remarks>Derived classes must override this method to get asynchronous behaviour. Otherwise it will
+        /// execute synchronously, returning an already-completed task. Asynchronous behaviour is also not available when the
+        /// constructor was passed an instance of type derived from <see cref="BinaryReader"/>.</remarks>
+        public override Task<DateTime?> ReadAsDateTimeAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return _asyncReader == null ? base.ReadAsDateTimeAsync(cancellationToken) : DoReadAsDateTimeAsync(cancellationToken);
+        }
+
+        private async Task<DateTime?> DoReadAsDateTimeAsync(CancellationToken cancellationToken)
+        {
+            switch (await GetContentTokenAsync(cancellationToken).ConfigureAwait(false))
+            {
+                case JsonToken.None:
+                case JsonToken.Null:
+                case JsonToken.EndArray:
+                    return null;
+                case JsonToken.Date:
+#if HAVE_DATE_TIME_OFFSET
+                    if (Value is DateTimeOffset)
+                    {
+                        SetToken(JsonToken.Date, ((DateTimeOffset)Value).DateTime, false);
+                    }
+#endif
+
+                    return (DateTime)Value;
+                case JsonToken.String:
+                    return ReadDateTimeString((string)Value);
```

Comment:
Another needed method. Again, if we had `ReadCurrentTokenAsDateTime()` it would give good re-use and be relatively clean in what it exposed.